modules/apps/item-selector/item-selector-impl/src/main/java/com/liferay/item/selector/internal/provider/GroupItemSelectorProviderImpl.java

```diff
@@ -73,10 +75,22 @@ public List<Group> getGroups(
 			).build();
 
 		try {
-			return _filterGroups(
-				_groupLocalService.search(
-					companyId, _classNameIds, keywords, groupParams, start, end,
-					null));
+			List<Group> groups = _groupLocalService.search(
+				companyId, _classNameIds, keywords, groupParams, start, end,
+				null);
+
+			List<Group> filteredGroups = _filterGroups(groups);
+
+			if (filteredGroups.size() < groups.size()) {
+				filteredGroups = ListUtil.subList(
+					_filterGroups(
+						_groupLocalService.search(
+							companyId, _classNameIds, keywords, groupParams,
+							QueryUtil.ALL_POS, QueryUtil.ALL_POS, null)),
+					start, end);
+			}
+
+			return filteredGroups;
```

Comment:
This will return duplicated entries when paginating. E.g. ask for groups (0, 10), where the 10th is not visible by the user, will return 0, 1, ..., 9, 11 (i.e. include the first visible group in the next page). When you navigate to the next page, the first element will be included again: 11, 12, 13, ...

So, we're exchanging one bug for another one. If there's no general solution, I'd prefer to accept the current behavior as a limitation.

Reply:
the problem is that it could happen that all Sites are filtered out from a page, so not only no Sites will be available, but it won't be possible to navigate between pages as the page selection links will disappear if there are no results on the page.
I noticed the duplication effect as well, but for me and presumably, for the customer, this would be a smaller issue than rendering the Sites unavailable.
Please let me know what you think.

Reply:
Ok, I see.

Could you avoid reading the whole Group table in a list? Instead, could you get groups page by page until you get `end - start` viewable elements?

Reply:
@adolfopa I created the alternative solution where the groups are queried from the subsequent pages instead of all at once. 
https://github.com/IstvanD/liferay-portal/commit/df03a492a8ad00b9e33d6dd8d66fe6af886e64ff
However, the problem with this approach is that the groups still won't align with the pagination segments, because they are calculated **before** the filtering takes place.
Let's say, we have groups 1-10, and we are displaying 4 items per page,
1-4 are filtered out, so we pull 5-8 from the next page.
Then, we navigate to page 2, where groups 5-8 are shown, but we cannot navigate further to page 3 to see groups 9-10, because the pagination algorithm calculates only 2 pages based on the number of filtered groups, which is 6.

In contrast, when segmenting the pages across the already filtered groups, on page 1 we get groups 5-8, and on page 2 groups 9-10.

Reply:
What worries me about your solution is that (in the worst case) you'll need to load the whole table into memory every time someone paginates through the group list. I think the fix can become a bigger issue than the original bug.

Reply:
Should we go with the other solution then?

Reply:
The alternative solution looks better to me, but I'm not sure about all this. By looking at the GroupServiceImpl class, the same problem is there. Maybe this fix (or similar) should go there instead of in the display context? Ideally, we should simply call groupService.getGroups and call it a day (the permission check is slightly different, but I think the extra checks we're doing here may not be necessary).

Can you add the alternative logic to GroupService and send it to core infra? Otherwise, the same problem will appear when paginating in other places where groups are displayed.

Reply:
@adolfopa the only other places (as I know of) where pagination for Groups is used are where the Administrator lists them or they are listed by membership. For example, a non-admin user can see the Sites in Control Panel with he has "Access in Control Panel" permission but only those he is member of.
In these situations, no follow-up filtering is taking place because the filtering is part of the query (for example, "list only the Sites I'm a member of").

Reply:
@adolfopa I have sent the updated solution.
https://github.com/liferay-lima/liferay-portal/pull/4252